src/action_runner/action_runner.cpp

```diff
@@ -33,6 +37,84 @@ int uninstall_msi_action()
     return 0;
 }
 
+namespace fs = std::filesystem;
+
+std::optional<fs::path> copy_self_to_temp_dir()
+{
+    std::error_code error;
+    auto dst_path = fs::temp_directory_path() / "action_runner.exe";
+    fs::copy_file(get_module_filename(), dst_path, fs::copy_options::overwrite_existing, error);
+    if (error)
+    {
+        return std::nullopt;
+    }
+    return std::move(dst_path);
+}
+
+bool install_new_version_stage_1(const bool must_restart = false)
+{
+    std::optional<fs::path> installer;
+    for (auto path : fs::directory_iterator{ updating::get_pending_updates_path() })
+    {
+        if (path.path().native().find(updating::installer_filename_pattern) != std::wstring::npos)
+        {
+            installer.emplace(std::move(path));
+            break;
+        }
+    }
+    if (!installer)
+    {
+        return false;
+    }
+
+    if (auto copy_in_temp = copy_self_to_temp_dir())
+    {
+        // detect if PT was running
+        const bool launch_powertoys = must_restart || FindWindowW(pt_tray_icon_window_class, nullptr) != nullptr;
+
+        std::wstring arguments{ L"-update_now_stage_2 \"" };
+        arguments += installer->c_str();
+        arguments += L"\" \"";
+        arguments += get_module_folderpath();
+        arguments += L"\" ";
+        arguments += launch_powertoys ? L"restart" : L"dont_start";
```

Comment:
shouldn't these args be constants since they are being referenced in other parts of code to prevent mistypes?

Reply:
done!